### PR TITLE
Jennifer Brabson PR | IC Week 2

### PR DESCRIPTION
**Pull request by JBrabson**
Made it through 4!

Used the extra time for additional functionality (ie: over capacity message), as well as helper methods (making a goal to write cleaner - aka no more 'end, end, end, end, end's... Bob witnessed the ugly). Can't promise they're good, but they work!

Thanks for helping me learn.

### CODE REVIEW

**Comment by JBrabson** [comment]
Also... while creeping to see if this has been graded... noticed I COMPLETELY missed hoarded. Just totally did not see that in the test. My eyes. Sabotage. Sooooooo the chances of a 4 are growing wings, flying away and out the window, I know... :(

File: lib/product.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class Product
```
Comment (robertguturing) [inline lib/product.rb:1]: It looks like you are missing the method implementation for `hoard` and `is_hoarded?` here, please read the readme a bit more closely next time 🙈

File: lib/shopping_cart.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
+class ShoppingCart
+  attr_reader :name,
+              :capacity,
+              :products
+
+  def initialize(name, capacity)
+    @name = name
+    @capacity = capacity.to_i
+    @products = []
+    @full = false
```
Comment (robertguturing) [inline lib/shopping_cart.rb:10]: I don't hate the idea of an instance variable for `@full` here but it also seems we like don't need it. We should only have instance variables for values we intend to change, and are not dynamically calculated. This instance variable does not meet either of those criteria as it only being used once one line 34.

File: lib/shopping_cart.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    }
+  end
+
+  def total_number_of_products
+    @products.sum do |product|
+      product.quantity
+    end
+  end
+
+  def is_full?
```
Comment (robertguturing) [inline lib/shopping_cart.rb:30]: We can simply this to `total_number_of_products > @capacity` and ditch the if else statment as this the code will always return a boolean value

File: lib/shopping_cart.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+  def percent_over
+    ((decimal_occ * 100) -100).to_i
+  end
+
+  def decimal_occ
+    total_number_of_products / @capacity.to_f
+  end
+
+  def sorted_products_by_quantity
```
Comment (robertguturing) [inline lib/shopping_cart.rb:62]: This is correct, but be careful the iteration asks us to name this method `products_by_quantity`

File: lib/shopping_cart.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+  def details
+    cart_details = {
+      name: @name,
+      capacity: @capacity
+    }
+  end
+
+  def total_number_of_products
+    @products.sum do |product|
```
Comment (robertguturing) [inline lib/shopping_cart.rb:25]: nice use of sum here

**Review by robertguturing — COMMENTED** [review]
Score : 3.5

I like the progress I'm seeing this inning already. This is a strong IC, just some notes about paying close attention to what the README is asking for, but nice work!

**Comment by JBrabson** [comment]
Thanks Bob!  Was a little pulled between IC and my dog yesterday (I think she's got something). So my eyes were crossed and I definitely missed some things. Frustrated with myself on that, but really appreciate the feedback and will watch for those next time. Grazie!!!

**Pull request closed without merging**